### PR TITLE
feat: add jsdom driver

### PR DESCRIPTION
**Pull request by Kikobeats**
Hello,

This PR introduces jsdom driver. I decided to create this driver because I want to wappalyzer passing all the necessary things as parameters

```js
const { url, html, headers } = await getHTML('https://kikobeats.com')
const wappalyzer = new Wappalyzer(url, { html, headers })
```

### CODE REVIEW

File: src/wappalyzer.js [hunk 1]
```diff
@@ -607,7 +607,7 @@ class Wappalyzer {
           headerName = headerName.toLowerCase();
 
           if (headerName in headers) {
-            headers[headerName].forEach((headerValue) => {
+            [].concat(headers[headerName]).compact(Boolean).forEach((headerValue) => {
```
Comment (Kikobeats) [inline src/wappalyzer.js]: @AliasIO not sure why the header value needs to be an array.

Normally, headers is an object like this:

```js
const headers = {
  date: 'Sat, 22 Feb 2020 16:50:20 GMT',
  server: 'cloudflare'
}
```

File: src/drivers/npm/package.json [hunk 2]
```diff
@@ -24,6 +24,8 @@
     "wappalyzer": "./cli.js"
   },
   "dependencies": {
+    "jsdom": "~16.2.0",
+    "tough-cookie": "~3.0.1",
     "zombie": "^6.1.2"
```
Comment (Kikobeats) [inline src/drivers/npm/package.json]: I recommend you ship every driver in a different wappalyzer, something like:

-  @wappalyzer/jsdom
-  @wappalyzer/puppeteer
-  @wappalyzer/zombie

Using this approach then you only install what you need

File: src/drivers/npm/browsers/jsdom.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+const uniq = arr => [...new Set(arr)]
+
+const parseCookie = str => Cookie.parse(str).toJSON()
+
+const getCookies = str =>
+  []
+    .concat(str)
+    .map(parseCookie)
+    .map(({ key: name, ...props }) => ({ name, ...props }))
```
Comment (Kikobeats) [inline src/drivers/npm/browsers/jsdom.js]: @AliasIO not sure why your implementation is using `name` instead of `key`

**Comment by AliasIO** [comment]
The Zombie.js driver is built on jsdom, is there a benefit to using this?

File: src/wappalyzer.js (hunk 1 above)
Comment (AliasIO) [reply src/wappalyzer.js]: It's to handle duplicate headers.

File: src/drivers/npm/package.json (hunk 2 above)
Comment (AliasIO) [reply src/drivers/npm/package.json]: Yes, I might do this at some point.

**Comment by Kikobeats** [comment]
I never used before, but looks like zombie performs a network call for getting the html content:
https://github.com/assaf/zombie#browservisiturl-options-callback

while in this implementation, all the things are passed via parameters, no internal network call will be performed.

In fact, if zombie is using jsdom as a dependency that's a good thing since the dependency will be deduped.

**Comment by Kikobeats** [comment]
@AliasIO 

The point with this driver is to not do an internal fetch of the data.

I feel this PR goal could be achieved adding support to pass your own data:

- at puppeter: if `html` is passed, set it using [page.setContent](https://github.com/puppeteer/puppeteer/blob/v2.1.1/docs/api.md#pagesetcontenthtml-options) and don't performing the goto call.

- at zombie: if `html` is passed, set is using [browser.load](https://github.com/assaf/zombie#browserloadhtml-callback), skipping `browser.visit` action.

That could be a more reasonable approach since I really don't think more driver can be added.

Thoughts?

**Comment by AliasIO** [comment]
I prefer the idea of separate packages, rather than officially supporting multiple browsers. You should already be able to create you own browser package and pull in Wappalyzer as a dependency. When I find the time I'll split the Zombie.js and Puppeteer browsers into their own package.

**Comment by Kikobeats** [comment]
Closed, I created https://github.com/Kikobeats/simple-wappalyzer instead 🙂

**Pull request closed without merging**